crates/net/eth-wire/src/multiplex.rs

```diff
@@ -487,19 +488,17 @@ where
                 }
             }
 
-            // advance all satellites
+            // advance all satellites once
             for idx in (0..this.inner.protocols.len()).rev() {
                 let mut proto = this.inner.protocols.swap_remove(idx);
-                loop {
-                    match proto.poll_next_unpin(cx) {
-                        Poll::Ready(Some(msg)) => {
-                            this.inner.out_buffer.push_back(msg);
-                        }
-                        Poll::Ready(None) => return Poll::Ready(None),
-                        Poll::Pending => {
-                            this.inner.protocols.push(proto);
-                            break
-                        }
+                match proto.poll_next_unpin(cx) {
+                    Poll::Ready(Some(msg)) => {
+                        this.inner.out_buffer.push_back(msg);
+                    }
+                    Poll::Ready(None) => return Poll::Ready(None),
+                    Poll::Pending => {
+                        this.inner.protocols.push(proto);
+                        break
```

Comment:
this break will terminate the for loop now, but we still want to advance the remaining streams, right?

perhaps we should rewrite `protocols: Vec<ProtocolStream>` as 
`FuturesUnordered<StreamFuture<ProtocolStream>`

see also SelectAll impl 

https://docs.rs/futures-util/0.3.30/src/futures_util/stream/select_all.rs.html#93

Reply:
absolutely. I also found some more code that impedes us from prioritising the primary stream, fixed it in the latest commit.